File: lib/sports/mlb/batterRankings.ts
/**
 * Position-aware batter rankings — the batter-side equivalent of
 * pitcherRankings.ts's role-classified pools, but split by primary position
 * (C/1B/2B/3B/SS/OF/DH) rather than starter/closer/reliever, since position
 * is what actually groups "who else should this hitter's contact quality be
 * compared against" the way a pitcher's role does. Same idea otherwise: every
 * batter gets ranked against the whole qualified pool (`overallRank`) and
 * against just their position group (`positionRank`), plus a composite
 * 0-100 score blending the four Statcast quality-of-contact metrics.
 *
 * Unlike pitcherRankings.ts, there's no traditional-stat merge here (no
 * AVG/OBP/SLG) — this is purely the Statcast side, the same four metrics
 * PlayerDetail's Quality of Contact card and the team-level rollup already
 * use. A composite blending traditional batting stats in too is a real
 * future extension, not part of this.
 */

import { readSnapshotCache, writeSnapshotCache } from '../../db/client';
import { getLeagueBatterSeasonRows, type RawBatterSeasonLine } from './statsapi';
import {
  getSeasonStatcastBatterRates,
  rankBatterStatcast,
  type BatterStatcastKey,
  type StatcastBatterRates,
} from './savant';

/** LF/CF/RF pool together — any one outfield spot alone is too small a pool to rank meaningfully against. Anything not listed here (a rare/legacy abbreviation) falls back to its own raw code rather than being silently dropped — see `positionGroupFor`. */
const POSITION_GROUP: Record<string, string> = {
  C: 'C',
  '1B': '1B',
  '2B': '2B',
  '3B': '3B',
  SS: 'SS',
  LF: 'OF',
  CF: 'OF',
  RF: 'OF',
  OF: 'OF',
  DH: 'DH',
};

function positionGroupFor(raw: string): string {
  return POSITION_GROUP[raw] ?? raw;
}

/**
 * Traditional season rates/counts — a separate, simpler ranker from the
 * Statcast composite above: no gating thresholds, no position split, all six
 * higher-is-better. Rides the same `getLeagueBatterSeasonRows` fetch and the
 * same 24h cache as the Statcast pool below rather than a new fetch/cache of
 * its own.
 */
export type BatterTraditionalKey = 'avg' | 'obp' | 'slg' | 'ops' | 'homeRuns' | 'rbi';

export const BATTER_TRADITIONAL_RANK_KEYS: ReadonlyArray<{ key: BatterTraditionalKey; label: string; decimals: number }> = [
  { key: 'avg', label: 'AVG', decimals: 3 },
  { key: 'obp', label: 'OBP', decimals: 3 },
  { key: 'slg', label: 'SLG', decimals: 3 },
  { key: 'ops', label: 'OPS', decimals: 3 },
  { key: 'homeRuns', label: 'HR', decimals: 0 },
  { key: 'rbi', label: 'RBI', decimals: 0 },
];

/** 1 = best, among every batter in `rows` that has a value for that stat. */
function rankBattersTraditional(rows: RawBatterSeasonLine[]): Map<number, Partial<Record<BatterTraditionalKey, number>>> {
  const ranks = new Map<number, Partial<Record<BatterTraditionalKey, number>>>(rows.map((r) => [r.personId, {}]));
  for (const { key } of BATTER_TRADITIONAL_RANK_KEYS) {
    const withValue = rows
      .filter((r) => r[key] != null)
      .sort((a, b) => (b[key] as number) - (a[key] as number));
    withValue.forEach((r, i) => {
      ranks.get(r.personId)![key] = i + 1;
    });
  }
  return ranks;
}

export interface RankedBatter {
  personId: number;
  fullName: string;
  teamId?: number;
  /** Grouped position (see `POSITION_GROUP`), not the raw StatsAPI abbreviation. */
  position: string;
  values: Partial<Record<BatterStatcastKey, number>>;
  overallRanks: Partial<Record<BatterStatcastKey, number>>;
  positionRanks: Partial<Record<BatterStatcastKey, number>>;
  /** Size of the whole qualified-batter pool this season (not position-scoped). */
  poolSize: number;
  /** Size of just this batter's position group. */
  positionPoolSize: number;
  /** 0-100, higher is better, against the whole pool — null when there aren't enough ranked stats to compute one. */
  composite: number | null;
  /** 1 = best against the whole pool. Null alongside a null composite. */
  overallRank: number | null;
  /** Same composite math, but scored only against this batter's position group. */
  positionComposite: number | null;
  /** 1 = best within this batter's position group. */
  positionRank: number | null;
  /** Traditional AVG/OBP/SLG/OPS/HR/RBI — see `BATTER_TRADITIONAL_RANK_KEYS`. Not position-scoped, not part of the Statcast composite. */
  traditionalValues: Partial<Record<BatterTraditionalKey, number>>;
  traditionalRanks: Partial<Record<BatterTraditionalKey, number>>;
  traditionalPoolSize: number;
}

export interface BatterRankings {
  season: number;
  computedAt: string;
  batters: RankedBatter[];
}

/**
 * Hand-set v1 weights — same "principled, not yet fit against real outcomes"
 * posture as pitcherRankings.ts's own weights. Barrel%/hard-hit% (batted-ball
 * authority) carry the most weight since they're the most direct read on
 * contact quality; exit velo overlaps with both so gets less independent
 * weight; whiff% (contact skill, not quality-of-contact) rounds it out.
 */
const COMPOSITE_WEIGHTS: Record<BatterStatcastKey, number> = {
  barrelPct: 0.3,
  hardHitPct: 0.3,
  exitVelo: 0.2,
  whiffPct: 0.2,
};

function compositeScore(ranks: Partial<Record<BatterStatcastKey, number>>, poolSize: number): number | null {
  if (poolSize <= 1) return null;
  // barrelPct/exitVelo/hardHitPct are gated together by the same battedBalls
  // threshold (MIN_BATTED_BALLS_FOR_QUALITY_RATE, savant.ts) — a batter has
  // all three or none. whiffPct alone (real swings, too few batted balls) is
  // a real number but too thin a profile to stand in for "quality of
  // contact": naive renormalization would let a 60-pitch September call-up's
  // lone elite whiff% outrank a full-season elite hitter on 100% weight
  // instead of the 20% it's actually worth. Requiring at least one of the
  // batted-ball trio keeps a composite from being computed on that alone.
  if (ranks.barrelPct == null && ranks.exitVelo == null && ranks.hardHitPct == null) return null;
  let weightedSum = 0;
  let weightUsed = 0;
  for (const [key, weight] of Object.entries(COMPOSITE_WEIGHTS) as [BatterStatcastKey, number][]) {
    const rank = ranks[key];
    if (rank == null) continue;
    // rank 1 (best) -> 100th percentile; rank poolSize (worst) -> 0th.
    const percentile = 100 * (1 - (rank - 1) / (poolSize - 1));
    weightedSum += percentile * weight;
    weightUsed += weight;
  }
  if (weightUsed === 0) return null;
  return weightedSum / weightUsed;
}

function buildPool(
  rows: RawBatterSeasonLine[],
  rates: Map<number, StatcastBatterRates>,
): { ranked: RankedBatter[]; ranksByPersonId: Map<number, Partial<Record<BatterStatcastKey, number>>>; poolSize: number } {
  const scopedRates = new Map<number, StatcastBatterRates>();
  for (const row of rows) {
    const rate = rates.get(row.personId);
    if (rate) scopedRates.set(row.personId, rate);
  }
  const ranks = rankBatterStatcast(scopedRates);
  const poolSize = scopedRates.size;

  const ranked = rows
    .filter((row) => scopedRates.has(row.personId))
    .map((row): RankedBatter => {
      const values = scopedRates.get(row.personId)!;
      const personRanks = ranks.get(row.personId) ?? {};
      const composite = compositeScore(personRanks, poolSize);
      return {
        personId: row.personId,
        fullName: row.fullName,
        teamId: row.teamId,
        position: positionGroupFor(row.position),
        values,
        overallRanks: personRanks,
        positionRanks: {},
        poolSize,
        positionPoolSize: 0,
        composite,
        overallRank: null,
        positionComposite: null,
        positionRank: null,
        traditionalValues: {},
        traditionalRanks: {},
        traditionalPoolSize: 0,
      };
    });

  return { ranked, ranksByPersonId: ranks, poolSize };
}

const CACHE_TTL_MS = 24 * 60 * 60_000;

function cacheKey(season: number): string {
  return `mlb:batter-rankings:${season}`;
}

/** Position-and-overall-ranked, composite-scored batter pool — 24h cached so a normal page load never pays for a Statcast/StatsAPI refresh. `forceRefresh` bypasses the cache (used by a manual "refresh now" trigger). */
export async function getBatterRankings(season: number, forceRefresh = false): Promise<BatterRankings> {
  if (!forceRefresh) {
    const cached = await readSnapshotCache(cacheKey(season));
    if (cached) {
      const age = Date.now() - Date.parse(cached.fetchedAt);
      if (age < CACHE_TTL_MS) {
        return JSON.parse(cached.payload) as BatterRankings;
      }
    }
  }

  const [rows, rates] = await Promise.all([getLeagueBatterSeasonRows(season), getSeasonStatcastBatterRates(season)]);

  // Overall pool first — every qualified batter, whatever their position.
  const overall = buildPool(rows, rates);
  const overallByPersonId = new Map(overall.ranked.map((b) => [b.personId, b]));

  // Traditional AVG/OBP/SLG/OPS/HR/RBI — ranked against every batter with a
  // season row (not gated to the Statcast-qualified pool above, so a batter
  // who's played enough to have real season totals but hasn't cleared the
  // Statcast batted-ball thresholds still gets these).
  const traditionalPoolSize = rows.length;
  const traditionalRanksByPersonId = rankBattersTraditional(rows);
  for (const row of rows) {
    const batter = overallByPersonId.get(row.personId);
    if (!batter) continue;
    batter.traditionalValues = {
      avg: row.avg,
      obp: row.obp,
      slg: row.slg,
      ops: row.ops,
      homeRuns: row.homeRuns,
      rbi: row.rbi,
    };
    batter.traditionalRanks = traditionalRanksByPersonId.get(row.personId) ?? {};
    batter.traditionalPoolSize = traditionalPoolSize;
  }

  // Composite ties (common — several batters landing on the exact same
  // ordinal rank for most of their four component stats) break on pitches
  // seen, the closest proxy for playing time this pipeline has — same
  // "bigger, more trustworthy sample wins an identical score" reasoning as
  // pitcherRankings.ts's innings-pitched tiebreak.
  const sampleSizeOf = (personId: number) => rates.get(personId)?.pitchSampleSize ?? 0;
  overall.ranked.sort((a, b) => {
    const composite = (b.composite ?? -1) - (a.composite ?? -1);
    if (composite !== 0) return composite;
    return sampleSizeOf(b.personId) - sampleSizeOf(a.personId);
  });
  overall.ranked.forEach((b, i) => {
    b.overallRank = b.composite != null ? i + 1 : null;
  });

  // Position pools — same rows/rates, grouped and re-ranked within each group.
  const rowsByPosition = new Map<string, RawBatterSeasonLine[]>();
  for (const row of rows) {
    const group = positionGroupFor(row.position);
    const bucket = rowsByPosition.get(group) ?? [];
    bucket.push(row);
    rowsByPosition.set(group, bucket);
  }

  for (const positionRows of rowsByPosition.values()) {
    const positionPool = buildPool(positionRows, rates);
    positionPool.ranked.sort((a, b) => {
      const composite = (b.composite ?? -1) - (a.composite ?? -1);
      if (composite !== 0) return composite;
      return sampleSizeOf(b.personId) - sampleSizeOf(a.personId);
    });
    positionPool.ranked.forEach((posBatter, i) => {
      const overallBatter = overallByPersonId.get(posBatter.personId);
      if (!overallBatter) return;
      overallBatter.positionRanks = posBatter.overallRanks;
      overallBatter.positionPoolSize = positionPool.poolSize;
      overallBatter.positionComposite = posBatter.composite;
      overallBatter.positionRank = posBatter.composite != null ? i + 1 : null;
    });
  }

  const result: BatterRankings = { season, computedAt: new Date().toISOString(), batters: overall.ranked };
  await writeSnapshotCache(cacheKey(season), JSON.stringify(result));
  return result;
}

/**
 * Same bundle, but never recomputes — reads whatever's cached, however
 * stale, and returns null if nothing's been computed yet this season. Safe
 * to call from the live snapshot-build path, same reasoning as
 * `getCachedPitcherRoleRankings` (pitcherRankings.ts): `getBatterRankings`
 * itself can fall through to a live Statcast/StatsAPI fetch on a stale/cold
 * cache, which is fine for a manually-triggered refresh but not for a
 * function every scan request goes through.
 */
export async function getCachedBatterRankings(season: number): Promise<BatterRankings | null> {
  const cached = await readSnapshotCache(cacheKey(season));
  if (!cached) return null;
  try {
    return JSON.parse(cached.payload) as BatterRankings;
  } catch {
    return null;
  }
}

File: python-odds-service/src/predict/batter_rankings.py
"""Direct port of lib/sports/mlb/batterRankings.ts — not a
reimplementation.

Position-aware batter rankings — every batter ranked against the whole
qualified pool (overall_rank) and against just their position group
(position_rank), plus a composite 0-100 score blending the four Statcast
quality-of-contact metrics. Diagnostics-only, same scope note as
pitcher_rankings.py — not part of the live matchup/candidate pipeline.

Cache shape matches TS's BatterRankings/RankedBatter interfaces field-
for-field (camelCase JSON keys), same cache-sharing intent as
pitcher_rankings.py/savant.py.
"""
import json
from dataclasses import dataclass, field
from datetime import datetime, timezone

import httpx

import db
from predict import savant, statsapi

# LF/CF/RF pool together — any one outfield spot alone is too small a
# pool to rank meaningfully against. Anything not listed here falls back
# to its own raw code rather than being silently dropped.
POSITION_GROUP: dict[str, str] = {
    "C": "C",
    "1B": "1B",
    "2B": "2B",
    "3B": "3B",
    "SS": "SS",
    "LF": "OF",
    "CF": "OF",
    "RF": "OF",
    "OF": "OF",
    "DH": "DH",
}


def _position_group_for(raw: str) -> str:
    return POSITION_GROUP.get(raw, raw)


# Traditional season rates/counts — a separate, simpler ranker from the
# Statcast composite: no gating thresholds, no position split, all six
# higher-is-better.
BATTER_TRADITIONAL_RANK_KEYS: list[tuple[str, str, int]] = [
    ("avg", "AVG", 3),
    ("obp", "OBP", 3),
    ("slg", "SLG", 3),
    ("ops", "OPS", 3),
    ("homeRuns", "HR", 0),
    ("rbi", "RBI", 0),
]

_TRADITIONAL_ATTR = {"avg": "avg", "obp": "obp", "slg": "slg", "ops": "ops", "homeRuns": "home_runs", "rbi": "rbi"}


def _rank_batters_traditional(rows: list[statsapi.RawBatterSeasonLine]) -> dict[int, dict[str, int]]:
    """1 = best, among every batter in `rows` that has a value for that stat."""
    ranks: dict[int, dict[str, int]] = {r.person_id: {} for r in rows}
    for key, _label, _decimals in BATTER_TRADITIONAL_RANK_KEYS:
        attr = _TRADITIONAL_ATTR[key]
        with_value = [r for r in rows if getattr(r, attr) is not None]
        with_value.sort(key=lambda r: getattr(r, attr), reverse=True)
        for i, r in enumerate(with_value):
            ranks[r.person_id][key] = i + 1
    return ranks


# Hand-set v1 weights — barrel%/hard-hit% (batted-ball authority) carry
# the most weight since they're the most direct read on contact quality;
# exit velo overlaps with both so gets less independent weight; whiff%
# (contact skill, not quality-of-contact) rounds it out.
#
# Keys are savant.py's own rank-dict keys (snake_case — its
# rank_batter_statcast output, built in an earlier phase of this port,
# uses barrel_pct/exit_velo/hard_hit_pct/whiff_pct, not TS's camelCase
# BatterStatcastKey spelling). The JSON cache serialization below
# translates to camelCase at the boundary so the on-disk shape still
# matches TS's BatterStatcastKey exactly.
COMPOSITE_WEIGHTS: dict[str, float] = {"barrel_pct": 0.3, "hard_hit_pct": 0.3, "exit_velo": 0.2, "whiff_pct": 0.2}


def _composite_score(ranks: dict[str, int], pool_size: int) -> float | None:
    if pool_size <= 1:
        return None
    # barrel_pct/exit_velo/hard_hit_pct are gated together by the same
    # battedBalls threshold (savant.py) — a batter has all three or none.
    # whiff_pct alone is too thin a profile to stand in for "quality of
    # contact" — requiring at least one of the batted-ball trio keeps a
    # composite from being computed on that alone.
    if ranks.get("barrel_pct") is None and ranks.get("exit_velo") is None and ranks.get("hard_hit_pct") is None:
        return None
    weighted_sum = 0.0
    weight_used = 0.0
    for key, weight in COMPOSITE_WEIGHTS.items():
        rank = ranks.get(key)
        if rank is None:
            continue
        percentile = 100 * (1 - (rank - 1) / (pool_size - 1))
        weighted_sum += percentile * weight
        weight_used += weight
    if weight_used == 0:
        return None
    return weighted_sum / weight_used


@dataclass
class RankedBatter:
    person_id: int
    full_name: str
    team_id: int | None
    position: str
    values: dict = field(default_factory=dict)
    overall_ranks: dict = field(default_factory=dict)
    position_ranks: dict = field(default_factory=dict)
    pool_size: int = 0
    position_pool_size: int = 0
    composite: float | None = None
    overall_rank: int | None = None
    position_composite: float | None = None
    position_rank: int | None = None
    traditional_values: dict = field(default_factory=dict)
    traditional_ranks: dict = field(default_factory=dict)
    traditional_pool_size: int = 0


def _build_pool(rows: list[statsapi.RawBatterSeasonLine], rates: dict[int, savant.StatcastRates]) -> tuple[list[RankedBatter], dict[int, dict[str, int]], int]:
    scoped_rates = {row.person_id: rates[row.person_id] for row in rows if row.person_id in rates}
    ranks = savant.rank_batter_statcast(scoped_rates)
    pool_size = len(scoped_rates)

    ranked: list[RankedBatter] = []
    for row in rows:
        if row.person_id not in scoped_rates:
            continue
        values = scoped_rates[row.person_id]
        person_ranks = ranks.get(row.person_id, {})
        composite = _composite_score(person_ranks, pool_size)
        ranked.append(
            RankedBatter(
                person_id=row.person_id,
                full_name=row.full_name,
                team_id=row.team_id,
                position=_position_group_for(row.position),
                values={"barrelPct": values.barrel_pct, "exitVelo": values.exit_velo, "hardHitPct": values.hard_hit_pct, "whiffPct": values.whiff_pct},
                overall_ranks=person_ranks,
                pool_size=pool_size,
                composite=composite,
            )
        )
    return ranked, ranks, pool_size


@dataclass
class BatterRankings:
    season: int
    computed_at: str
    batters: list[RankedBatter]


CACHE_TTL_MS = 24 * 60 * 60_000


def _cache_key(season: int) -> str:
    return f"mlb:batter-rankings:{season}"


async def get_batter_rankings(client: httpx.AsyncClient, season: int, force_refresh: bool = False) -> BatterRankings:
    """Position-and-overall-ranked, composite-scored batter pool — 24h
    cached so a normal page load never pays for a Statcast/StatsAPI
    refresh. `force_refresh` bypasses the cache."""
    if not force_refresh:
        cached = await db.read_snapshot_with_age(_cache_key(season))
        if cached is not None:
            payload, age_seconds = cached
            if age_seconds * 1000 < CACHE_TTL_MS:
                return _rankings_from_json(json.loads(payload))

    rows = await statsapi.get_league_batter_season_rows(client, season)
    rates = await savant.get_season_statcast_batter_rates(client, season)

    # Overall pool first — every qualified batter, whatever their position.
    overall_ranked, _overall_ranks, _overall_pool_size = _build_pool(rows, rates)
    overall_by_person_id = {b.person_id: b for b in overall_ranked}

    # Traditional AVG/OBP/SLG/OPS/HR/RBI — ranked against every batter
    # with a season row (not gated to the Statcast-qualified pool above).
    traditional_pool_size = len(rows)
    traditional_ranks_by_person_id = _rank_batters_traditional(rows)
    for row in rows:
        batter = overall_by_person_id.get(row.person_id)
        if batter is None:
            continue
        batter.traditional_values = {"avg": row.avg, "obp": row.obp, "slg": row.slg, "ops": row.ops, "homeRuns": row.home_runs, "rbi": row.rbi}
        batter.traditional_ranks = traditional_ranks_by_person_id.get(row.person_id, {})
        batter.traditional_pool_size = traditional_pool_size

    # Composite ties break on pitches seen, the closest proxy for playing
    # time this pipeline has — same "bigger, more trustworthy sample wins
    # an identical score" reasoning as pitcher_rankings.py's IP tiebreak.
    def sample_size_of(person_id: int) -> float:
        r = rates.get(person_id)
        return r.pitch_sample_size if r else 0

    overall_ranked.sort(key=lambda b: (-(b.composite if b.composite is not None else -1), -sample_size_of(b.person_id)))
    for i, b in enumerate(overall_ranked):
        b.overall_rank = i + 1 if b.composite is not None else None

    # Position pools — same rows/rates, grouped and re-ranked within each group.
    rows_by_position: dict[str, list[statsapi.RawBatterSeasonLine]] = {}
    for row in rows:
        group = _position_group_for(row.position)
        rows_by_position.setdefault(group, []).append(row)

    for position_rows in rows_by_position.values():
        position_ranked, _pos_ranks, position_pool_size = _build_pool(position_rows, rates)
        position_ranked.sort(key=lambda b: (-(b.composite if b.composite is not None else -1), -sample_size_of(b.person_id)))
        for i, pos_batter in enumerate(position_ranked):
            overall_batter = overall_by_person_id.get(pos_batter.person_id)
            if overall_batter is None:
                continue
            overall_batter.position_ranks = pos_batter.overall_ranks
            overall_batter.position_pool_size = position_pool_size
            overall_batter.position_composite = pos_batter.composite
            overall_batter.position_rank = i + 1 if pos_batter.composite is not None else None

    result = BatterRankings(season=season, computed_at=datetime.now(timezone.utc).isoformat(), batters=overall_ranked)
    await db.write_snapshot(_cache_key(season), json.dumps(_rankings_to_json(result)))
    return result


async def get_cached_batter_rankings(season: int) -> BatterRankings | None:
    """Same bundle, but never recomputes — reads whatever's cached,
    however stale, and returns None if nothing's been computed yet this
    season. Safe to call from a live request path."""
    payload = await db.read_snapshot(_cache_key(season))
    if payload is None:
        return None
    try:
        return _rankings_from_json(json.loads(payload))
    except (json.JSONDecodeError, KeyError, TypeError):
        return None


# savant.py's rank-dict keys (snake_case) -> TS's BatterStatcastKey
# spelling (camelCase) — see COMPOSITE_WEIGHTS's comment above for why
# these differ internally.
_RANK_KEY_TO_CAMEL = {"barrel_pct": "barrelPct", "exit_velo": "exitVelo", "hard_hit_pct": "hardHitPct", "whiff_pct": "whiffPct"}
_RANK_KEY_FROM_CAMEL = {v: k for k, v in _RANK_KEY_TO_CAMEL.items()}


def _ranks_to_json(ranks: dict) -> dict:
    return {_RANK_KEY_TO_CAMEL.get(k, k): v for k, v in ranks.items()}


def _ranks_from_json(d: dict) -> dict:
    return {_RANK_KEY_FROM_CAMEL.get(k, k): v for k, v in d.items()}


def _batter_to_json(b: RankedBatter) -> dict:
    return {
        "personId": b.person_id,
        "fullName": b.full_name,
        "teamId": b.team_id,
        "position": b.position,
        "values": b.values,
        "overallRanks": _ranks_to_json(b.overall_ranks),
        "positionRanks": _ranks_to_json(b.position_ranks),
        "poolSize": b.pool_size,
        "positionPoolSize": b.position_pool_size,
        "composite": b.composite,
        "overallRank": b.overall_rank,
        "positionComposite": b.position_composite,
        "positionRank": b.position_rank,
        "traditionalValues": b.traditional_values,
        "traditionalRanks": b.traditional_ranks,
        "traditionalPoolSize": b.traditional_pool_size,
    }


def _batter_from_json(d: dict) -> RankedBatter:
    return RankedBatter(
        person_id=d["personId"],
        full_name=d["fullName"],
        team_id=d.get("teamId"),
        position=d["position"],
        values=d.get("values") or {},
        overall_ranks=_ranks_from_json(d.get("overallRanks") or {}),
        position_ranks=_ranks_from_json(d.get("positionRanks") or {}),
        pool_size=d.get("poolSize") or 0,
        position_pool_size=d.get("positionPoolSize") or 0,
        composite=d.get("composite"),
        overall_rank=d.get("overallRank"),
        position_composite=d.get("positionComposite"),
        position_rank=d.get("positionRank"),
        traditional_values=d.get("traditionalValues") or {},
        traditional_ranks=d.get("traditionalRanks") or {},
        traditional_pool_size=d.get("traditionalPoolSize") or 0,
    )


def _rankings_to_json(r: BatterRankings) -> dict:
    return {"season": r.season, "computedAt": r.computed_at, "batters": [_batter_to_json(b) for b in r.batters]}


def _rankings_from_json(d: dict) -> BatterRankings:
    return BatterRankings(season=d["season"], computed_at=d["computedAt"], batters=[_batter_from_json(b) for b in d.get("batters") or []])
